### DISCUSSION TITLE
concourse keys rotation

### DISCUSSION THREAD

**Discussion by gowrisankar22** [concourse/concourse#5876] [Help & Support]
Hello All,

The concourse is using **host-keys**, **worker-keys** and  **session-signing-keys**  and rotation of all theses keys require recreation of web and worker node which will cause the downtime. If there any recommended way we can rotate this without downtime or its not possible. Also, how often it is recommended to rotate?

BRs, Gowrisankar

**Comment by crsimmons** [comment]
I believe you have to restart the web and worker processes in order to rotate the keys. That doesn't necessarily have to cause downtime though. If you're running more than one web and worker nodes you should be able to roll them one by one assuming your deployment mechanism allows for it. For example I know that BOSH and Helm bot allow you to do rolling updates.

**Reply by richo6030** [reply]
Hi @crsimmons , as I'm also working in the direction of implementing a rotation for the same keys but in a Bosh environment I was wondering - what is the best approach to rotate these keys? Again the goal is not to cause any downtime, but does Concourse support any blue/green approach for this or it is recommended that we use Bosh vm strategy to adopt the rotation without causing disruptions in the system?

**Reply by crsimmons** [reply]
I'm not aware of a blue-green feature within Concourse. Generally Concourse is agnostic to how it is deployed.

With BOSH I would just use the [update block in the manifest](https://bosh.io/docs/manifest-v2/#update) to configure max_in_flight for the web and worker instance groups to ensure at least one is always up. If you've got a load balancer in front of your webs there shouldn't be downtime there. IIRC the drain script on workers will wait for running jobs to finish.
